libbeat/features/features.go

```diff
@@ -0,0 +1,86 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package features
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+var (
+	flags = fflags{}
+)
+
+type fflags struct {
+	mu sync.RWMutex
+
+	fqdnEnabled bool
+}
+
+// UpdateFromProto updates the feature flags configuration. If f is nil UpdateFromProto is no-op.
+func UpdateFromProto(f *proto.Features) {
+	if f == nil {
+		logp.L().Debug("feature flags are nil, ignoring them")
+		return
+	}
+
+	if f.Fqdn == nil {
+		f.Fqdn = &proto.FQDNFeature{}
+	}
+
+	flags.mu.Lock()
+	defer flags.mu.Unlock()
+	flags.fqdnEnabled = f.Fqdn.Enabled
+}
+
+// UpdateFromConfig updates the feature flags configuration. If c is nil UpdateFromProto is no-op.
+func UpdateFromConfig(c *conf.C) error {
+	if c == nil {
+		logp.L().Debug("feature flags are nil, ignoring them")
+		return nil
+	}
+
+	type cfg struct {
+		Features struct {
+			FQDN *conf.C `json:"fqdn" yaml:"fqdn" config:"fqdn"`
+		} `json:"features" yaml:"features" config:"features"`
+	}
+
+	parsedFlags := cfg{}
+	if err := c.Unpack(&parsedFlags); err != nil {
+		return fmt.Errorf("could not Unpack features config: %w", err)
+	}
+
+	flags.mu.Lock()
+	defer flags.mu.Unlock()
+	flags.fqdnEnabled = parsedFlags.Features.FQDN.Enabled()
```

Comment:
can we hit nil ref with FQDN or features?

Reply:
We can't hit a nil ref with `parsedFlags.Features` because it's a struct (not pointer).  But good catch on `parsedFlags.Features.FQDN`.  I will add a guard.  Thanks.

Reply:
Actually, if `parsedFlags.Features.FQDN` is `nil`, `parsedFlags.Features.FQDN.Enabled()` will still handle it and return `false`:

https://github.com/elastic/elastic-agent-libs/blob/0ba2aa0ce1ac3a7a3dd511854e1cc65463784bb4/config/config.go#L246-L248

So I think we're good here? Or do you think we should add a guard anyway?